### PR TITLE
add cast(value as JSON) support.

### PR DESCRIPTION
**Pull request by hicqu**

### CODE REVIEW

**Comment by hicqu** [comment]
I will add test cases after #3388 merged.

File: util/types/datum_test.go [hunk 1]
```diff
@@ -184,6 +184,52 @@ func (ts *testTypeConvertSuite) TestToFloat32(c *C) {
...
+		{NewStringDatum("\"hello, 世界\""), `"hello, 世界"`, true},
+		{NewStringDatum("[1, 2, 3]"), `[1, 2, 3]`, true},
+		{NewStringDatum("{}"), `{}`, true},
+		{NewIntDatum(1), `true`, true},
+		{mustParseTimeIntoDatum("2011-11-10 11:11:11", mysql.TypeTimestamp, 0), `"2011-11-10 11:11:11"`, true},
+
+		{NewStringDatum("hello, 世界"), "", false},
+	}
+	for _, tt := range tests {
+		obtain, err := tt.datum.ConvertTo(sc, ft)
```
Comment (shenli) [inline util/types/datum_test.go]: c.Assert(tt.success, Equals, err == nil) and remove the if-else

File: util/types/datum.go [hunk 2]
```diff
@@ -1195,11 +1195,31 @@ func (d *Datum) convertToMysqlJSON(sc *variable.StatementContext, target *FieldT
 		var j json.JSON
 		if j, err = json.ParseFromString(d.GetString()); err == nil {
 			ret.SetMysqlJSON(j)
+		} else {
+			// If parse fail, create a JSON with the string primitive.
```
Comment (shenli) [inline util/types/datum.go]: Any test case cover this?

Comment (hicqu) [inline util/types/datum.go]: it's a wrong logic. I will remove this.

File: util/types/datum_test.go (hunk 1 above)
Comment (hicqu) [inline util/types/datum_test.go]: I prefer keep the if-else, because I can compare obtained and expected datum only if success == true.

**Comment by hicqu** [comment]
@XuHuaiyu @coocood , PTAL, thanks!

File: util/types/datum_test.go [hunk 3]
```diff
@@ -184,6 +184,53 @@ func (ts *testTypeConvertSuite) TestToFloat32(c *C) {
...
+		{NewIntDatum(1), `1.0`, true},
+		{NewFloat64Datum(2), `2`, true},
+		{NewStringDatum("\"hello, 世界\""), `"hello, 世界"`, true},
+		{NewStringDatum("[1, 2, 3]"), `[1, 2, 3]`, true},
+		{NewStringDatum("{}"), `{}`, true},
+		{NewIntDatum(1), `true`, true},
+		{mustParseTimeIntoDatum("2011-11-10 11:11:11", mysql.TypeTimestamp, 0), `"2011-11-10 11:11:11"`, true},
+
+		// can not parse JSON from this string, so error occurs.
+		{NewStringDatum("hello, 世界"), "", false},
```
Comment (XuHuaiyu) [inline util/types/datum_test.go]: why here?

File: util/types/datum.go [hunk 4]
```diff
@@ -1196,10 +1196,27 @@ func (d *Datum) convertToMysqlJSON(sc *variable.StatementContext, target *FieldT
...
+		}
 	case KindMysqlJSON:
 		ret = *d
 	default:
-		return invalidConv(d, target.Tp)
+		var s string
+		if s, err = d.ToString(); err == nil {
+			// TODO: fix precision of MysqlTime. For example,
+			// On MySQL 5.7 CAST(NOW() AS JSON) -> "2011-11-11 11:11:11.111111",
+			// But now we can only return "2011-11-11 11:11:11".
```
Comment (XuHuaiyu) [inline util/types/datum.go]: why we lost the precision here?

Comment (hicqu) [inline util/types/datum.go]: it's because string format in `Time.ToString`.

File: util/types/datum_test.go (hunk 3 above)
Comment (hicqu) [inline util/types/datum_test.go]: If we convert string to JSON, we should parse that string, not use that string as JSON primitive. So here if an invalid JSON string is given, the `ConvertTo` should fail.

File: util/types/datum_test.go [hunk 5]
```diff
@@ -184,6 +184,53 @@ func (ts *testTypeConvertSuite) TestToFloat32(c *C) {
...
+		expected string
+		success  bool
+	}{
+		{NewIntDatum(1), `1.0`, true},
+		{NewFloat64Datum(2), `2`, true},
+		{NewStringDatum("\"hello, 世界\""), `"hello, 世界"`, true},
+		{NewStringDatum("[1, 2, 3]"), `[1, 2, 3]`, true},
+		{NewStringDatum("{}"), `{}`, true},
+		{NewIntDatum(1), `true`, true},
+		{mustParseTimeIntoDatum("2011-11-10 11:11:11.111111", mysql.TypeTimestamp, 6), `"2011-11-10 11:11:11.111111"`, true},
```
Comment (coocood) [inline util/types/datum_test.go]: Any failed cases?

**Comment by coocood** [comment]
There is no SQL level test case.

**Comment by coocood** [comment]
And parser test

**Comment by hicqu** [comment]
test cases for parser and SQL level are added. PTAL, thanks.

File: util/types/datum.go [hunk 6]
```diff
@@ -1196,10 +1196,27 @@ func (d *Datum) convertToMysqlJSON(sc *variable.StatementContext, target *FieldT
...
 		}
+	case KindInt64, KindUint64:
+		i64 := d.GetInt64()
+		ret.SetMysqlJSON(json.CreateJSON(i64))
+	case KindFloat32, KindFloat64:
+		f64 := d.GetFloat64()
+		ret.SetMysqlJSON(json.CreateJSON(f64))
+	case KindMysqlDecimal:
+		var f64 float64
+		if f64, err = d.GetMysqlDecimal().ToFloat64(); err == nil {
```
Comment (tiancaiamao) [inline util/types/datum.go]: what happen when err != nil? such as truncated.

File: util/types/datum_test.go [hunk 7]
```diff
@@ -184,6 +184,53 @@ func (ts *testTypeConvertSuite) TestToFloat32(c *C) {
 	c.Assert(converted.GetFloat64(), Equals, datum.GetFloat64())
 }
 
+// mustParseTime is similar to ParseTime but panic if any error occurs.
```
Comment (tiancaiamao) [inline util/types/datum_test.go]: Inconsistent function name and comment name.

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**